// orbiter-chaincore/src/decorators/cache.decorator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cacheDecorator = void 0;
const keyv_1 = __importDefault(require("keyv"));
const keyvFile_1 = __importDefault(require("../utils/keyvFile"));
const cacheDecorator = (file) => {
    return (target, propertyKey) => {
        const cache = new keyv_1.default({
            store: new keyvFile_1.default({
                filename: `cache/${file}`,
                expiredCheckDelay: 999999 * 24 * 3600 * 1000,
                writeDelay: 100,
                encode: JSON.stringify,
                decode: JSON.parse, // deserialize function
            }),
        });
        Object.defineProperty(target, propertyKey, {
            value: cache,
            writable: false,
            enumerable: true,
            configurable: true,
        });
    };
};
exports.cacheDecorator = cacheDecorator;
