// orbiter-chaincore/src/types/transaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = exports.TransactionStatus = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
var TransactionStatus;
(function (TransactionStatus) {
    TransactionStatus[TransactionStatus["PENDING"] = 0] = "PENDING";
    TransactionStatus[TransactionStatus["COMPLETE"] = 1] = "COMPLETE";
    TransactionStatus[TransactionStatus["Fail"] = 2] = "Fail";
})(TransactionStatus = exports.TransactionStatus || (exports.TransactionStatus = {}));
class Transaction {
    /**
     * @Deprecated
     * field name error
     */
    constructor(opts) {
        this.chainId = String(opts.chainId || "");
        this.hash = String(opts.hash || "");
        this.nonce = Number(opts.nonce || 0);
        this.blockHash = String(opts.blockHash || "");
        this.blockNumber = Number(opts.blockNumber || 0);
        this.transactionIndex = Number(opts.transactionIndex || 0);
        this.from = opts.from || "";
        this.to = opts.to || "";
        this.value = opts.value || new bignumber_js_1.default(0);
        this.symbol = opts.symbol;
        this.gasPrice = Number(opts.gasPrice || 0);
        this.gas = Number(opts.gas || 0);
        this.input = opts.input || "";
        this.status = opts.status || TransactionStatus.PENDING;
        this.tokenAddress = String(opts.tokenAddress || "");
        this.timestamp = Number(opts.timestamp);
        this.fee = String(opts.fee || 0);
        this.feeToken = String(opts.feeToken || "");
        this.extra = opts.extra;
        this.source = String(opts.source || "");
        this.confirmations = Number(opts.confirmations || 1);
        this.toChainId = String(opts.toChainId || "");
        this.receipt = opts.receipt;
    }
}
exports.Transaction = Transaction;
