// orbiter-chaincore/src/watch/dydx.watch.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DydxWatch = void 0;
const dayjs_1 = __importDefault(require("dayjs"));
const core_1 = require("../utils/core");
const base_watch_1 = __importDefault(require("./base.watch"));
class DydxWatch extends base_watch_1.default {
    constructor() {
        super(...arguments);
        this.minConfirmations = 0;
    }
    getApiFilter(address) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = {
                limit: 50,
                createdBeforeOrAt: (0, dayjs_1.default)().toISOString(),
            };
            const cursor = yield this.apiScanCursor(address);
            if (cursor) {
                params.createdBeforeOrAt = String((0, dayjs_1.default)(cursor.timestamp).toISOString());
            }
            return params;
        });
    }
    apiWatchNewTransaction(address) {
        return __awaiter(this, void 0, void 0, function* () {
            if ((0, core_1.isEmpty)(this.chain.chainConfig.api.key)) {
                return [];
            }
            const filter = yield this.getApiFilter(address);
            const response = yield this.chain.getTransactions(address, filter);
            return response.txlist;
        });
    }
    replayBlock(_start, _end, _changeBlock) {
        throw new Error("Method not implemented.");
    }
}
exports.DydxWatch = DydxWatch;
