// orbiter-chaincore/src/watch/immutableX.watch.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImmutableXWatch = void 0;
const dayjs_1 = __importDefault(require("dayjs"));
const base_watch_1 = __importDefault(require("./base.watch"));
class ImmutableXWatch extends base_watch_1.default {
    constructor() {
        super(...arguments);
        this.minConfirmations = 0;
    }
    getApiFilter(address) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = {
                page_size: 100,
                direction: "desc",
            };
            yield this.apiScanCursor(address);
            return params;
        });
    }
    apiWatchNewTransaction(address) {
        return __awaiter(this, void 0, void 0, function* () {
            const filter = yield this.getApiFilter(address);
            const response = yield this.chain.getTransactions(address, filter);
            const cursor = yield this.apiScanCursor(address);
            const prevTxTime = (cursor === null || cursor === void 0 ? void 0 : cursor.timestamp) || (0, dayjs_1.default)().unix();
            return response.txlist.filter(tx => tx.timestamp > prevTxTime);
        });
    }
    replayBlock(_start, _end, _changeBlock) {
        throw new Error("Method not implemented.");
    }
}
exports.ImmutableXWatch = ImmutableXWatch;
