// orbiter-chaincore/src/watch/zksync.watch.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZKSyncWatch = void 0;
const core_1 = require("../utils/core");
const base_watch_1 = __importDefault(require("./base.watch"));
class ZKSyncWatch extends base_watch_1.default {
    constructor() {
        super(...arguments);
        this.minConfirmations = 0;
    }
    getApiFilter(address) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = {
                from: "latest",
                limit: 100,
                direction: "newer",
            };
            const cursor = yield this.apiScanCursor(address);
            if (cursor && cursor.hash) {
                params.from = String(cursor.hash);
            }
            return params;
        });
    }
    apiWatchNewTransaction(address) {
        return __awaiter(this, void 0, void 0, function* () {
            const filter = yield this.getApiFilter(address);
            const response = yield this.chain.getTransactions(address, filter);
            return response.txlist.filter(row => (0, core_1.equals)(row.extra["type"], "Transfer"));
        });
    }
    replayBlock(_start, _end, _changeBlock) {
        throw new Error("Method not implemented.");
    }
}
exports.ZKSyncWatch = ZKSyncWatch;
