// orbiter-chaincore/src/watch/zksync2.watcht.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZKSync2Watch = void 0;
const evm_watch_1 = __importDefault(require("./evm.watch"));
class ZKSync2Watch extends evm_watch_1.default {
    constructor() {
        super(...arguments);
        this.minConfirmations = 3;
        this.feeAddress = "0xde03a0B5963f75f1C8485B355fF6D30f3093BDE7".toLowerCase();
    }
    getApiFilter(address) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = {
                address,
                sort: "asc",
                page: 1,
                offset: 0,
            };
            const cursor = yield this.apiScanCursor(address);
            if (cursor) {
                params.startblock = Number(cursor.blockNumber) + 1;
            }
            return params;
        });
    }
    pushBefore(address, txList) {
        return __awaiter(this, void 0, void 0, function* () {
            txList = txList.filter(tx => tx.to.toLowerCase() != this.feeAddress);
            return { address, txList };
        });
    }
    apiWatchNewTransaction(address) {
        return __awaiter(this, void 0, void 0, function* () {
            const filter = yield this.getApiFilter(address);
            const response = yield this.chain.getTokenTransactions(address, null, filter);
            response.txlist = response.txlist.filter(row => row.to.toLowerCase() !== this.feeAddress);
            return response.txlist;
        });
    }
}
exports.ZKSync2Watch = ZKSync2Watch;
